// wisepaas-datahub-edge-nodejs-sdk/model/edge/EdgeConfig.js
'use strict';
class EdgeConfig {
  constructor () {
    this.Scada = new ScadaConfig();
    return this;
  }
}

class NodeConfig {
  constructor () {
    this.name = '';
    this.description = '';
    this.deviceList = [];
    return this;
  }
}

class ScadaConfig {
  constructor () {
    this.Id = '';
    this.Name = '';
    this.DeviceType = 0;
    this.Description = '';
    this.HeartBeat = 1;
    this.PortNumber = 1;
    this.BackupDeviceId = 1;
    this.AnalogTagList = [];
    this.DiscreteTagList = [];
    this.TextTagList = [];
    return this;
  }
}

class TagConfig {
  constructor () {
    this.Name = '';
    this.ReadOnly = false;
    this.Description = '';
    this.ArraySize = 2;

  }
}

class AnalogTagConfig extends TagConfig {
  constructor () {
    super();
    // this.name = 'ATag';
    this.Description = '';
    this.NeedLog = true;
    this.ArraySize = 2;
    this.SpanHigh = 1000;
    this.SpanLow = 0;
    this.EngineerUnit = '';
    this.DisplayFormat = "4.2";
    this.AlarmEnable = true;
    this.HHPriority = 0;
    this.HHAlarmLimit = 0;
    this.HighPriority = 0;
    this.LLPriority = 0;
    this.LowPriority = 0;
    this.LowAlarmLimit = 0;
    this.LLAlarmLimit = 0;

    return this;
  }
}

class DiscreteTagConfig extends TagConfig {
  constructor () {
    super();
    // this.name = 'DTag';
    this.description = '';
    this.readOnly = false;
    this.arraySize = 0;
    this.state0 = '0';
    this.state1 = '1';
    this.state2 = '';
    this.state3 = '';
    this.state4 = '';
    this.state5 = '';
    this.state6 = '';
    this.state7 = '';
    return this;
  }
}

class TextTagConfig extends TagConfig {
  constructor () {
    super();
    return this;
  }
}

module.exports = {
  EdgeConfig,
  NodeConfig,
  ScadaConfig,
  AnalogTagConfig,
  DiscreteTagConfig,
  TextTagConfig
};
